arraycontext/context.py

```diff
@@ -123,10 +113,55 @@
 from pytools.tag import Tag
 
 
-DeviceArray = Any
-DeviceScalar = Any
+# {{{ typing
+
 _ScalarLike = Union[int, float, complex, np.generic]
 
+try:
+    from typing import Protocol
+except ImportError:
+    from typing_extensions import Protocol                  # type: ignore[misc]
+
+
+class DeviceArray(Protocol):
```

Comment:
Suppose this is our last change to discuss the name. Is `DeviceArray` the best we can do? It's kind of a poor fit for many of the use cases. (I know there's a precedent, but we *could* change/deprecate this before we plaster this name absolutely everywhere.)

How about just `Array`? Or `ContextArray` (:vomiting_face:)?

Reply:
Hm, I'm not too bothered by `DeviceArray`. That's what JAX calls its CPU/GPU array class and `Device` makes sense in the `pyopencl` case too. I imagine it's a worse fit for the lazy case?

I wouldn't feel bad about just calling it `Array` either (as a component of an `ArrayContainer`). @kaushikcfd Thoughts or suggestions for better naming?

Reply:
`Array` or `ArrayT` sound better, compared to the other options here :). (numpy arrays are also not quite "device arrays")

Reply:
Renamed to `Array` in https://github.com/inducer/arraycontext/compare/af65ccc1c8b6edbb2d454757a726ef0cd3484431..e7a0fd3972ba3d794ffd0d656b3e736fa42d1392